arangod/GeneralServer/RestHandler.cpp

```diff
@@ -403,9 +403,13 @@ void RestHandler::handleExceptionPtr(std::exception_ptr eptr) noexcept try {
   // can do here to signal this problem.
 }
 
-void RestHandler::runHandlerStateMachine() {
+auto RestHandler::runHandlerStateMachine(bool isWakeup) -> HandlerState {
   std::lock_guard lock{_executionMutex};
-  TRI_ASSERT(_callback);
+  TRI_ASSERT(_sendResponseCallback);
+
+  if (isWakeup && _state != HandlerState::PAUSED) {
+    return _state;
+  }
```

Comment:
Can we move this into the `wakeupHandler`? Then we don't need this `isWakeup` parameter at all. The `wakeupHandler` would need the lock to check the flag, but can release it before calling runHandlerStateMachine. Alternatively I would remove the lock and rename this to `runHandlerStateMachineLocked` and add another `runHandlerStateMachine` that just acquires the lock and calls `runHandlerStateMachineLocked`. The `wakeupHandler` could then also just call that one.